feat: connector v1.3 review

D-Nice opened the pull request.

gskapka committed:
standard fxn formatting

...which also makes the visibility specifiers & fxn modifiers easier to
see.

gskapka committed:
adds `getPrice` overloads & makes them `view`

Now users can query prices for their (or others') contracts!

gskapka committed:
switch to explicit `uint256`

gskapka committed:
standardizes fxn formatting

gskapka committed:
adds `getPrice` overloads to connector interface

gskapka committed:
fix * spacing

gskapka committed:
fixes linting isues

gskapka committed:
adds gasPriceBumping feature to 0.5API

gskapka committed:
adds gasPriceBumping feature to 0.4.25API

gskapka committed:
adds cached-query feature to 0.4.25 API

gskapka committed:
adds cached-query feature to 0.4.25 API

gskapka committed:
adds parameterized & view getPrice fxns to 0.5 API

gskapka committed:
adds parameterized & view getPrice fxns to 0.4.25 API

gskapka committed:
add spec. for datasource string->byte conversion

...to both `0.4.25` & `0.5` version of the API.

gskapka committed:
updates README w/ currently maintained API versions

gskapka committed:
neatens natspec comments in 0.5 API

gskapka committed:
amends rebroadcast feat in 0.5 API

...and removes the now outdated `gasPriceBump` version

gskapka committed:
amends rebroadcast feat in 0.4.25 API

...and thus removes the now out-dated `gasPriceBump` nomenclature.

gskapka committed:
exposes token payment fxns in 0.4.25 API

gskapka committed:
exposes token payment fxns in 0.5 API

gskapka committed:
exposes token approval compostions in 0.4.25 API

gskapka committed:
exposes token approval compostions in 0.5 API

gskapka committed:
add missing datasource byte spec to 0.4.25 API

gskapka committed:
add missing datasource byte spec to 0.5 API

gskapka committed:
adds bytes datasource query overloads to 0.5 API

...and has to switch to lower level calls w/ a sprinkling of assembly to
get at the return values in order to deal with overloaded interfaces.

Pertinent issue here:
https://github.com/ethereum/solidity/issues/526#issuecomment-469271013i

gskapka committed:
switch to using `bytes1` explicitly in Oraclize Interface

gskapka committed:
reorder query overloads in 0.5API

...to get a neater separation of the queries from those with the dynamic
arguments.

gskapka committed:
adds query overloads w/ dynArgs & byte datasource to 0.5API

gskapka committed:
add byte-datasource overloads to interface for 0.4.25API

gskapka committed:
adds bytes version of Oraclize Interface to 0.4.25API

gskapka committed:
adds `bytes1` overloads of oraclize queries to 0.4.25API

gskapka committed:
adds `bytes1` datasource dynamic arg. query overloads to 0.4.25API

gskapka committed:
brings goerli comment in line with others

gskapka committed:
pass value to rebroadcast request in 0.4.25API

gskapka committed:
pass value to rebroadcast request in 0.5API

gskapka committed:
adds value param to cachedQuery fxn call in API 0.4.25

gskapka committed:
adds value param to cachedQuery fxn call in API 0.5

gskapka committed:
neaten commentary and format per natspec

gskapka committed:
unify param formatting in API 0.5

gskapka committed:
format interface fxns per new style guide in API 0.5

gskapka committed:
adds erc20 price getter to connector interface in 0.5 API

gskapka committed:
adds erc20 price getter to connector interface in 0.4.25 API

gskapka committed:
add token payment mapping to API 0.4.25

gskapka committed:
add token payment mapping to 0.5 API

gskapka committed:
adds `getERC20Price` overloads to 0.5 API

gskapka committed:
adds `getERC20Price` overloads to 0.4.25 API

D-Nice commented on oraclizeAPI_0.4.25.sol:
```diff
@@ -31,21 +31,55 @@ THE SOFTWARE.
-pragma solidity >= 0.4.22 < 0.5;// Incompatible compiler version... please select one stated within pragma solidity or use different oraclizeAPI version
+pragma solidity >= 0.4.22 < 0.5; // Incompatible compiler version... please select one stated within pragma solidity or use different oraclizeAPI version
+
+contract OraclizeIBytes {
```
comment explaining need for this segregation may be useful

gskapka replied:
Kk, agreed. Comment added.

D-Nice commented:
Regarding this, in the future, it may make more sense to do any stylistic refactors first, and merge them in, as it pollutes the code review, and can enter a review hypnosis where what I'm thinking are stylistic changes may be real code changes.

D-Nice commented on oraclizeAPI_0.4.25.sol:
```diff
@@ -303,23 +337,26 @@ contract usingOraclize {
         if(address(oraclize) != OAR.getAddress())
             oraclize = OraclizeI(OAR.getAddress());
-
+            oraclizeBytes = OraclizeIBytes(OAR.getAddress());
```
can save some ops by just doing `= address(oraclize)` here, no?

gskapka replied:
Ahh wait yes, because the line before ensures `oraclize` is set. Good spot!

D-Nice commented on oraclizeAPI_0.4.25.sol:
```diff
@@ -362,157 +399,1290 @@ contract usingOraclize {
 
-    function oraclize_getPrice(string datasource) oraclizeAPI internal returns (uint){
-        return oraclize.getPrice(datasource);
+    function __callback(
```
i believe this placeholder doesn't work in 0.5 as we want anymore, remember user reporting warning (I know this is 0.4.25 but mentioining here still).

gskapka replied:
So do both of the `__callback` functions need to disappear form the 0.5 version of the API?

D-Nice replied:
No we've decided to keep them with whatever was needed to silence them

D-Nice commented on oraclizeAPI_0.4.25.sol:
```diff
@@ -362,157 +399,1290 @@ contract usingOraclize {
+     * @notice oraclize_getPrice(...) overloads follow...
+     *
+     */
+    function oraclize_getPrice(
```
wouldn't it be useful to specify these as view?

gskapka replied:
Oops - remembered the less obvious bit of making it all view inside the interfaces but not here! Fixed!

D-Nice replied:
correction to my request, due to oraclizeAPI actually having the potential to modify state, we've dropped this.

D-Nice commented on oraclizeAPI_0.4.25.sol:
```diff
@@ -362,157 +399,1290 @@ contract usingOraclize {
+     * @notice oraclize_getPriceERC20(...) overloads follow...
+     *
+     */
+    function oraclize_getPriceERC20(
```
view?

gskapka replied:
These cannot be view due to `oraclizeAPI` modifiying state (as we've discussed elsewhere). Resolving convo.

D-Nice commented:
ping @gskapka

gskapka committed:
explain need for two interfaces in API 0.4.25

gskapka committed:
use extant address to save on opcodes when initting bytes interface

gskapka committed:
make getPrice overloads view in API 0.4.25

gskapka commented:
Agreed re the stylistic changes btw. Didn't mean to accidentally do both in tandem, but I did :/ Will ensure to separate the two concerns in future if both sorts of changes are needed. Have pushed all the fixes to the above comments.

D-Nice commented on oraclizeAPI_0.4.25.sol:
```diff
@@ -352,7 +352,7 @@ contract usingOraclize {
         if(address(oraclize) != OAR.getAddress())
             oraclize = OraclizeI(OAR.getAddress());
-            oraclizeBytes = OraclizeIBytes(OAR.getAddress());
+            oraclizeBytes = OraclizeIBytes(address(oraclize));
```
further can be saved, by pushing the OAR.getAddress() result somewhere on the stack first (save in a variable somewhere before the if statement) and re-using that across all 3 instances of OAR.getAddress here.

D-Nice replied:
Including in the coupon modifier that's still around.

gskapka replied:
Good spot, have optimized and push changes to this PR. 

Note that the coupon modifier doesn't exist in the 0.5 API (which is hopefully correct) and in the `0.4.25API` the coupon modifier only calls `OAR.getAddress()` once, so no savings to be gained there.

D-Nice commented on oraclizeAPI_0.5.sol:
```diff
@@ -1347,7 +1347,9 @@ contract usingOraclize {
         returns (bytes32 _id)
     {
-        uint256 price = oraclize.getPrice(_datasource);
+        uint256 price = oraclize.addressCustomPaymentToken(address(this)) == address(0)
```
I'm almost wondering if we should make another overload ~nightmare~ of oraclize_query(..., supportTokens = true) or oraclize_query_token?

I think these costs will be quite significant that it'll sacrifice optimizations, and push costs onto everyone.

gskapka replied:
Have discussed this elsewhere to resolving the conversation here.

D-Nice replied:
Sorry for the confusion on this and telling you to go this route first, I was initially under the impression there was just a ternary inclusion. You did also mention a storage read, which would've still been okay, as that would be 200 gas, and I considered we had 500/600 gas to play with. Of course, it's an external call involving a storage read, which can get quite expensive.

gskapka replied:
No worries, I was too vague in pointing it out initialy. No harm done, will have the changes on your (zenhub) desk by (your timezone's) morning.

D-Nice commented on oraclizeAPI_0.4.25.sol:
```diff
@@ -1080,7 +1080,9 @@ contract usingOraclize {
         returns (bytes32 id)
     {
-        uint price = oraclize.getPrice(datasource);
+        uint256 price = oraclize.addressCustomPaymentToken(address(this)) == address(0)
```
I'm definitely thinking we should have a separate set of these methods, for the case of contracts wanting to support ETH and token payments, for efficiency's sake and to not burden ETH only users. We are agnostic to which payment option is used.

gskapka replied:
Ibid last comment.

D-Nice reviewed (commented):
Additional points needing addressing

gskapka committed:
feat(0.5API): adds entire ERC20 interface

...since it's free w/r/t compile size if functions go unused.

gskapka committed:
feat(0.5API): adds entire ERC20 interface

gskapka committed:
optimize(0.5API): rm double lookup of connector address via OAR

gskapka committed:
optimize(0.4API): rm double lookup of connector address via OAR

gskapka committed:
feat(0.5API): adds `oraclize_token_query` overloads

... & also cleans up the natspec comments surrounding the various
overload types.

gskapka committed:
feat(0.4API): adds `oraclize_token_query` overloads

... & also update the natspec comments to make the API easier to
navigate.

gskapka committed:
fix(0.5API): fixes last remaining compiler warning re pure fxn

D-Nice committed:
fix(0.4API): revert getPrice views

reverting changes I requested for consistency with 0.5 and previous iterations

D-Nice committed:
refactor: improve specifier order consistency

D-Nice commented:
LGTM as of https://github.com/oraclize/ethereum-api/pull/88/commits/6a0b495dca8c91e69e098f719bccb22c98b11e3b

rest please review as well

gskapka commented:
Of your commits, I couldn't approve of the specifier order consistency more! Wohoo! 

So yeah, LGTM too. 

Have given @riccardopersiani an in-person overview of the changes because if you look at the diffs for the whole branch, they don't even load due to their size! It's given him a fighting chance at a review though & he's on it.

gskapka reviewed (dismissed):
Of your commits, I couldn't approve of the specifier order consistency more! Wohoo! 

So yeah, LGTM too. 

Have given @riccardopersiani an in-person overview of the changes because if you look at the diffs for the whole branch, they don't even load due to their size! It's given him a fighting chance at a review though & he's on it.

riccardopersiani commented on oraclizeAPI_0.4.25.sol:
```diff
@@ -756,164 +2159,2827 @@ contract usingOraclize {
+        dynargs[1] = args[1];
+        dynargs[2] = args[2];
+        dynargs[3] = args[3];
+        dynargs[4
```
Newbie question here... We are comparing `keccak256` because the string compare method would cost more?

gskapka replied:
Yup, as mentioned yesterday.

riccardopersiani commented on oraclizeAPI_0.4.25.sol:
```diff
@@ -1074,72 +5175,102 @@ contract usingOraclize {
-        for (uint256 i=0; i< n_random_bytes; i++) {
-            if (content[i] != prefix[i]) match_ = false;
+        require(_prefix.length == _nRandomBytes);
+        for (uint256 i = 0; i< _nRandomBytes; i++) {
```
i< need a space

gskapka replied:
`0.4.25` where stylistic changes weren't made ∴ ignoring.

riccardopersiani commented on oraclizeAPI_0.5.sol:
```diff
@@ -845,36 +5826,181 @@ contract usingOraclize {
+     * @notice Oraclize helper functions follow...
+     *
+     */
+    function parseAddr(string memory _a)
```
Where this function is used internally?

gskapka replied:
It's not, it's a helper method for end contracts to use should they need. Internal just means the non-using contracts who inherit this don't have it added to their bytecode.

riccardopersiani commented on oraclizeAPI_0.4.25.sol:
```diff
@@ -1029,7 +5120,13 @@ contract usingOraclize {
-    function oraclize_randomDS_proofVerify__sessionKeyValidity(bytes proof, uint sig2offset) internal returns (bool) {
+    function oraclize_randomDS_proofVerify__sessionKeyValidity(
+        bytes proof,
+        uint sig2offset
```
Don't know if it is worth to mention or correct every `uint` to `uint256`

gskapka replied:
Ignoring all `uints` in `0.4.25. Those in `0.5` are done. Ignoring remaining mentions.

riccardopersiani commented on oraclizeAPI_0.5.sol:
```diff
@@ -226,59 +543,100 @@ library CBOR {
+        pure
+    {
         if (_value >= 0) {
             encodeType(_buf, MAJOR_TYPE_INT, uint(_value));
```
Is it possible to cast to uint256?

D-Nice replied:
sure, but it's an external lib, and for easier import let's leave it as is.

riccardopersiani commented on oraclizeAPI_0.5.sol:
```diff
@@ -1,94 +1,389 @@
+    event Transfer(
+        address indexed _from,
+        address indexed _to,
+        uint tokens
```
uint256? follows 6 more

riccardopersiani commented on oraclizeAPI_0.5.sol:
```diff
@@ -984,24 +6168,41 @@ contract usingOraclize {
                 }
                 mint *= 10;
                 mint += uint(uint8(bresult[i])) - 48;
             } else if (uint(uint8(bresult[i])) == 46) {
```
There is bunch of uint that still cab converted to uint256 in the whole api

riccardopersiani commented on oraclizeAPI_0.4.25.sol:
```diff
@@ -366,155 +420,1286 @@ contract usingOraclize {
+
+    function oraclize_getPriceERC20(
+        byte _datasource,
+        address _address,
```
@gskapka was saying _address is to vague, maybe something more specific is needed

gskapka replied:
Change(s) made and functions much clearer now!

riccardopersiani reviewed (changes requested):
Review still need to be finished

riccardopersiani commented on oraclizeAPI_0.4.25.sol:
```diff
@@ -31,21 +31,71 @@ THE SOFTWARE.
     function randomDS_getSessionPubKeyHash() external constant returns(bytes32);
+    function getPrice(string memory _datasource) public view returns (uint256 _dsprice);
+    function getPrice(byte _datasource, address _address) public view returns (uint256 _dsprice);
+    function getPrice(byte _datasource, uint256 _gasLimit) public view returns (uint256 _dsprice);
```
As yesterday @gskapka was saying is better to use a more specific name for _address

gskapka replied:
`_address` elucidated (see below) & re your other question the function signatures would suggest that the first gets a price for a datasource and for a given address, and the second for a datasouce and a given gas limit.

gskapka committed:
chore(0.5API): use `uint256` explicitly where missing

gskapka committed:
chore(0.5API): elucidate address in `getPrice` fxns

gskapka committed:
chore(0.4.25API): elucidate address in `getPrice` fxns

gskapka commented on oraclizeAPI_0.4.25.sol:
```diff
@@ -31,21 +31,71 @@ THE SOFTWARE.
+}
 
 contract OraclizeI {
+
```
You were to ignore stylistic changes to `0.4.25` since it was decided not to update it since the work would be largely redundant.

gskapka commented on oraclizeAPI_0.4.25.sol:
```diff
@@ -285,9 +335,9 @@ End solidity-cborutils
-    uint constant month = 60*60*24*30;
+    uint constant day = 60 * 60 * 24;
+    uint constant week = 60 * 60 * 24 * 7;
+    uint constant month = 60 * 60 * 24 * 30;
```
Ibid comment above re stylistic changes to `0.4.25` w/r/t operator spacing ∴ won't fix.

Re `month_`, I don't see a gain from adding more weight to the API on deploy by adding more constants here since they exist for legacy/backwards-compatibility reasons.

gskapka commented on oraclizeAPI_0.4.25.sol:
```diff
@@ -303,23 +353,27 @@ contract usingOraclize {
     OraclizeAddrResolverI OAR;
```
Presumably legacy from the eth classic split where `id 2` was both eth classic's testnet id as well as eth's morden testnet?

Not sure if comment needed since this has never been brought up before and the curious could do their own research?

D-Nice replied:
this was back when morden was the sole testnet, so it waas referred to testnet generally, just as mainnet had several names along the lines of frontier, homestead etc... etc..

gskapka commented on oraclizeAPI_0.4.25.sol:
```diff
@@ -303,23 +353,27 @@ contract usingOraclize {
+        networkID; // silence the warning and remain backwards compatible
     }
     function oraclize_setNetwork() internal returns(bool){
         if (getCodeSize(0x1d3B2638a7cC9f2CB3D298A3DA7a90B67E5506ed)>0){ //mainnet
```
Ibid first comment re how you were to ignore all stylistic things w/r/t `0.4.25` since those changes were only made in `0.5` and non exist there.

gskapka commented on oraclizeAPI_0.4.25.sol:
```diff
@@ -366,155 +420,1286 @@ contract usingOraclize {
+    {
       myid; result; proof;
       oraclize_randomDS_args[bytes32(0)] = bytes32(0);
     }
```
Ignoring ∵ `0.4.25`. See prev. comment somewhere.

gskapka commented on oraclizeAPI_0.4.25.sol:
```diff
@@ -366,155 +420,1286 @@ contract usingOraclize {
     }
+    /**
+     *
+     * @notice oraclize_getPrice(...) overloads follow...
```
After elucidating the `_address` param via a rename I think the functions are much clearer and so we can avoid yet more comments.

gskapka commented on oraclizeAPI_0.4.25.sol:
```diff
@@ -366,155 +420,1286 @@ contract usingOraclize {
-    function oraclize_query(uint timestamp, string datasource, string arg, uint gaslimit) oraclizeAPI internal returns (bytes32 id){
+
+    function oraclize_query(
+        uint timestamp,
```
Pretty sure this would cost more gas since the EVM is optimized for 32 byte words and so more ops are needed to convert to/from when dealing with smaller sizes. Feel free to test.

D-Nice replied:
agree with greg, only makes sense for structs/storage.

gskapka commented:
Changes per @riccardopersiani review pushed & include: 

:heavy_check_mark: All 24 `uint` related things now solved with a better regex of `uint\(256\)\@!\(8\)\@!\(160\)\@!` in the find & replace. Note that this is in `0.5 API` _only_. 

:heavy_check_mark: `address _address` param in `getPrice` overloads now elucidated via renaming to `address _contractToQuery`.

All comments made re stylistic changes pertaining to `0.4.25` were ignored since the file itself should have been ignored in that regard during review.

D-Nice commented on oraclizeAPI_0.5.sol:
```diff
@@ -1013,16 +6214,30 @@ contract usingOraclize {
-            if ((uint(uint8(bresult[i])) >= 48) && (uint(uint8(bresult[i])) <= 57)) {
+        for (uint256 i = 0; i < bresult.length; i++) {
+            if (
+                (uint(uint8(bresult[i])) >= 48) &&
```
uint256 especially here, as @riccardopersiani pointed out would help this make more sense.

gskapka replied:
It tells me that you're looking at an outdated commit. All the `uint`s are now explicit.

riccardopersiani reviewed (approved):
LGTM

D-Nice commented:
In need of rebase (draft #94)

TODOs post-rebase 

- [ ] update copyright in rebased oraclize api
- [ ] update provableAPI with changes of this PR in single commit

D-Nice commented:
rebased and matches https://github.com/provable-things/ethereum-api/pull/94

D-Nice committed:
chore(copyright): update

D-Nice committed:
feat: update provable api to 1.3

also includes various fixes/consistency unto oraclize apis

D-Nice commented:
Only commit https://github.com/provable-things/ethereum-api/pull/88/commits/35d89331bc710ccacd4f733e1d348babdc39f758

should need review @riccardopersiani 

If there is anything questionable, please mention and request review from @gskapka in addition

D-Nice commented:
tests failing due to android backend issue, being resolved...

riccardopersiani reviewed (changes requested):
Added just very small notes. 

For the rest LGTM. Good job!

D-Nice committed:
refactor(style): code review suggestions

courtesy @riccardopersiani

Co-Authored-By: Riccardo Persiani <r.persiani92@gmail.com>

D-Nice commented:
resolved
